In [ ]:

# Kaggle Resource Spotify Songs 2024
# https://www.kaggle.com/datasets/nelgiriyewithana/most-streamed-spotify-songs-2024
# Kaggle Resource Spotify Songs 2023
# https://www.kaggle.com/datasets/nelgiriyewithana/top-spotify-songs-2023

# !pip install kaggle


In [ ]:
from google.colab import drive

import pandas as pd

spotify23_path = '/content/spotify-2023.csv'
spotify24_path = '/content/SpotifySongs2024.csv'



Spotify 2023

In [ ]:
df = pd.read_csv(spotify23_path, header = 1)
df.head()
df.columns = df.columns.str.strip()

df['released_year'] = pd.to_numeric(df['released_year'], errors='coerce')

df_2023 = df[df['released_year'] == 2023]

df_2023 = df_2023.rename(columns={
    "artist(s)_name": "artist_name",
    "in_spotify_charts": "peak_position",
    "streams": "stream_count",
    "bpm": "tempo",
    "danceability_%": "danceability",
    "valence_%": "valence",
    "energy_%": "energy",
    "acousticness_%": "acousticness",
    "instrumentalness_%": "instrumentalness",
    "speechiness_%": "speechiness",
    "liveness_%":"liveness"
})
df_2023.head()


track_name        artist_name  artist_count  \
0  Seven (feat. Latto) (Explicit Ver.)   Latto, Jung Kook             2   
1                                 LALA        Myke Towers             1   
2                              vampire     Olivia Rodrigo             1   
4                       WHERE SHE GOES          Bad Bunny             1   
5                             Sprinter  Dave, Central Cee             2   

   released_year  released_month  released_day  in_spotify_playlists  \
0           2023               7            14                   553   
1           2023               3            23                  1474   
2           2023               6            30                  1397   
4           2023               5            18                  3133   
5           2023               6             1                  2186   

   peak_position stream_count  in_apple_playlists  ...  tempo key   mode  \
0            147    141381703                  43  ...    125   B  Major   
1             48    133716286                  48  ...     92  C#  Major   
2            113    140003974                  94  ...    138   F  Major   
4             50    303236322                  84  ...    144   A  Minor   
5             91    183706234                  67  ...    141  C#  Major   

  danceability  valence energy acousticness  instrumentalness  liveness  \
0           80       89     83           31                 0         8   
1           71       61     74            7                 0        10   
2           51       32     53           17                 0        31   
4           65       23     80           14                63        11   
5           92       66     58           19                 0         8   

   speechiness  
0            4  
1            4  
2            6  
4            6  
5           24  

[5 rows x 24 columns]

In [ ]:
columns_to_drop = ["artist_count",
    "released_day",
    "in_spotify_playlists",
    "in_apple_playlists",
    "in_apple_charts",
    "deezer_in_playlists",
    "deezer_in_charts",
    "shazam_charts",]

df_2023 = df_2023.drop(columns = columns_to_drop, errors = "ignore")
df_2023.head(1).isna().any()

track_name             False
artist_name            False
released_year          False
released_month         False
peak_position          False
stream_count           False
in_deezer_playlists    False
in_deezer_charts       False
in_shazam_charts       False
tempo                  False
key                    False
mode                   False
danceability           False
valence                False
energy                 False
acousticness           False
instrumentalness       False
liveness               False
speechiness            False
dtype: bool

Spotify 2024

In [ ]:
df2 = pd.read_csv(spotify24_path, header=1)
df2.head()
df2['Release Date'] = pd.to_datetime(df2['Release Date'], errors='coerce')

# Filter rows where the year is 2024
df_2024 = df2[df2['Release Date'].dt.year == 2024]
df_2024 = df_2024.rename(columns={
    "Track": "track_name",
    "Artist": "artist_name",
    "Release Date": "release_date",
    "Explicit Track": "explicit",
    "Spotify Popularity": "track_popularity",
    "All Time Rank": "peak_position",
    "Spotify Streams": "stream_count"
})

df_2024.head()

track_name                    Album Name     artist_name  \
0         MILLION DOLLAR BABY  Million Dollar Baby - Single   Tommy Richman   
1                 Not Like Us                   Not Like Us  Kendrick Lamar   
2  i like the way you kiss me    I like the way you kiss me         Artemas   
4                     Houdini                       Houdini          Eminem   
6            Beautiful Things              Beautiful Things    Benson Boone   

  release_date          ISRC peak_position  Track Score stream_count  \
0   2024-04-26  QM24S2402528             1        725.4  390,470,936   
1   2024-05-04  USUG12400910             2        545.9  323,703,884   
2   2024-03-19  QZJ842400387             3        538.4  601,309,283   
4   2024-05-31  USUG12403398             5        423.3  107,034,922   
6   2024-01-18  USWB12307016             7        407.2  900,158,751   

  Spotify Playlist Count Spotify Playlist Reach  ...  SiriusXM Spins  \
0                 30,716            196,631,588  ...             684   
1                 28,113            174,597,137  ...               3   
2                 54,331            211,607,669  ...             536   
4                  7,223            151,469,874  ...               1   
6                 73,118            201,585,714  ...             429   

  Deezer Playlist Count Deezer Playlist Reach Amazon Playlist Count  \
0                  62.0            17,598,718                 114.0   
1                  67.0            10,422,430                 111.0   
2                 136.0            36,321,847                 172.0   
4                  82.0            17,660,624                 105.0   
6                 168.0            48,197,850                 154.0   

  Pandora Streams Pandora Track Stations Soundcloud Streams  Shazam Counts  \
0      18,004,655                 22,931          4,818,457      2,669,262   
1       7,780,028                 28,444          6,623,075      1,118,279   
2       5,022,621                  5,639          7,208,651      5,285,340   
4       4,493,884                  7,006            207,179        457,017   
6      65,447,476                 57,372                NaN      9,990,302   

  TIDAL Popularity explicit  
0              NaN        0  
1              NaN        1  
2              NaN        0  
4              NaN        1  
6              NaN        0  

[5 rows x 29 columns]

In [ ]:

df_2024 = df_2024.drop(columns = ["TIDAL Popularity", "Album Name", "ISRC"], errors = "ignore")

df_2024.head(1).isna().any()
df_2024.head()


track_name     artist_name release_date peak_position  \
0         MILLION DOLLAR BABY   Tommy Richman   2024-04-26             1   
1                 Not Like Us  Kendrick Lamar   2024-05-04             2   
2  i like the way you kiss me         Artemas   2024-03-19             3   
4                     Houdini          Eminem   2024-05-31             5   
6            Beautiful Things    Benson Boone   2024-01-18             7   

   Track Score stream_count Spotify Playlist Count Spotify Playlist Reach  \
0        725.4  390,470,936                 30,716            196,631,588   
1        545.9  323,703,884                 28,113            174,597,137   
2        538.4  601,309,283                 54,331            211,607,669   
4        423.3  107,034,922                  7,223            151,469,874   
6        407.2  900,158,751                 73,118            201,585,714   

   track_popularity YouTube Views  ... AirPlay Spins SiriusXM Spins  \
0              92.0    84,274,754  ...        40,975            684   
1              92.0   116,347,040  ...        40,778              3   
2              92.0   122,599,116  ...        74,333            536   
4              88.0    77,373,957  ...        12,185              1   
6              86.0   308,723,145  ...       383,478            429   

  Deezer Playlist Count Deezer Playlist Reach Amazon Playlist Count  \
0                  62.0            17,598,718                 114.0   
1                  67.0            10,422,430                 111.0   
2                 136.0            36,321,847                 172.0   
4                  82.0            17,660,624                 105.0   
6                 168.0            48,197,850                 154.0   

   Pandora Streams Pandora Track Stations Soundcloud Streams  Shazam Counts  \
0       18,004,655                 22,931          4,818,457      2,669,262   
1        7,780,028                 28,444          6,623,075      1,118,279   
2        5,022,621                  5,639          7,208,651      5,285,340   
4        4,493,884                  7,006            207,179        457,017   
6       65,447,476                 57,372                NaN      9,990,302   

  explicit  
0        0  
1        1  
2        0  
4        1  
6        0  

[5 rows x 26 columns]

Working with Spotify 2023


track_name	artist(s)_name	artist_count	year_published	released_month	released_day	spotify_in_play	spotify_in_chart	streams	apple_in_play	in_apple_charts	deezer_in_play	deezer_in_chart	shazam_charts	bpm	key	mode	danceability_	valence_	energy_	acousticness_	instrumentalness_	liveness_	speechiness_


MISSING ITEMS:
Song level:
duration_ms, explicit, track_popularity, peak_position,
Artist:
artist_popularity, artist_followers, artist_genres, past_hit_ratio
Lyrics:
sentiment_score, lyrical_complexity, language, loudness
Time:
season, day_of_week, trends window

REDUNDANT ITEMS:
artist_count, spotify_inplay, apple in play, deezer in play

Working with Spotify 2024

Track	Album Name	Artist	Release Date	ISRC	All Time Rank	Track Score	Spotify Streams	Spotify Playlist Count	Spotify Playlist Reach	Spotify Popularity	YouTube Views	YouTube Likes	TikTok Posts	TikTok Likes	TikTok Views	YouTube Playlist Reach	Apple Music Playlist Count	AirPlay Spins	SiriusXM Spins	Deezer Playlist Count	Deezer Playlist Reach	Amazon Playlist Count	Pandora Streams	Pandora Track Stations	Soundcloud Streams	Shazam Counts	TIDAL Popularity	Explicit Track

MISSING ITEMS:
danceability, energy, tempo/bmp, valence, loudness, acousticenss, instumentalness, speechiness, liveness

artist features:
artist_follwers, artsist_popularity, artist_enres, past_hit_ratio, track_popularity, peak_position

Lyrics:
sentiment_score, lyrical_complexity, language

SongLevel metadata:
duration_ms, key and mode

Time:
release_year, release_month

REDUNDANT ITEMS:
spotify playlist count, spotfiy playlist reach, youtube playlist reach, youtube views, aitplay spins, siriusxm spins, deezer playlist count, deezer playlist reach, amazon playlist count, pandiora streams, pandora track stations, tidal popularity

SECTION TO DO

Audio Features:
  danceability, energy, tempo, valence, loudness, acousticeness, instrumentalness, speechiness

Song level metadata:
  track_name, explicit or flagged, duration_ms, key, mode, release_date(maybe trends change within certain amount of days or monts),
Artist-level-info:
  artist_name, artist_popularity, artist_followers, artist_genres, past_hit_ratio
Performance history:
  track_popularity, peak_position, stream_count
Lyrics/nlp features:
  sentiment_score, lyrical_complexity, language
Time Features:
  release_year, release_month, season, day_of_week

TODO

CLEAN AND PROCESS THE DATA

remove or fix missing values convert dates into usefule features(month, season, etc) encode categorical features: artist name and genre scale numerical features: tempo, followers

FEATURE ENGINEERING

past_hit_ration: % of an artist's past songs that were hits artist_avg_popularity: mean popularity of all prior songs release_season: categorical;winter,summer use lyrics to calculate sentiment_score and,lyrical_complexity; include social metrics, spotify trends

Modeling

[split dataset by time] train 2023 songs validation early 2024 test: late 2024

[choose model type] baseline, treebased, neural networks

[train your model] fit on training data validate using the holdout 2024 set tune hyperparamets with gridsearhc or optuna

EVALUTAION & INTERPRETATION

[evaluate model] for classification: accuracy for regression: MAE for ranking: Precision@K

[Understand the Model] Feature importance use SHAP or LIME for deepre insight

PREDICTION TOOL

[Build Prediction Function] Create a function or script that allows inputs like predict_hit('artist': "Olivia Rodrigo", 'track': "new love") Outputs: predicted popularity: 85, hit porbability 88%

[EXTENSION] build an interface: StreamLit: interacive app, FastApi: serve the model via an API, Jupyter Notebook: for internal testing

FUTURE-Proofing

[Retrain Periodically]: Add social media data, playlist placemetns, collaborator success history

With a spotify Api

we can get most of information except: past_hit_ratio, peak_position, stream_count, sentiment_score, lyrical_complexity, language, trend_windows